working/0731 - horizontal inference/feature-specification.md

```diff
@@ -0,0 +1,290 @@
+```
+
+Today this doesn't work, because without the leading `int`, the downwards
+inference phase of type inference has no information with which to choose a
+preliminary type for the type parameter `T`, so the type context used for
+inference of the function literal `(a, b) => a < b ? b : a` is `? Function(?,
+int)`.  Hence, `a` gets assigned a static type of `Object?`, and this leads to a
+compile error at `a < b`.
+
+With horizontal inference, type inference first visits the `0`, and uses its
+static type to assign a preliminary type of `int` to the type parameter `T`.
+Then, when it proceeds to perform type inference on the function literal, the
+type context is `int Function(int, int)`.  This results in `a` being assigned a
+static type of `int`, which is what the user intends.
+
+## Type inference algorithm for invocations
+
+The full algorithm for type inference of invocations is expanded as follows.
+This algorithm is used for constructor invocations (including those in
+annotations, super-constructor calls, and constructor redirects), method
+invocations (including super calls), and invocations of function-typed
+expressions.
+
+1. Initial constraints: Create some constraints on type parameters by trying to
+   match the return type of the invocation target as a subtype of the incoming
+   type context.  (For a constructor invocation, the return type of the
+   invocation target is considered the raw uninstantiated type of the class
+   enclosing the constructor declaration.)
+
+2. Downwards inference: partially solve the set of type constraints accumulated
+   in step 1, to produce a preliminary mapping of type parameters to type
+   schemas.
+   
+3. Visit arguments (legacy): If experiment flag `inference-update-1` is not
+   enabled, recursively perform inference on all arguments to the invocation in
+   source order.  Obtain the type contexts for the recursive inference by
+   substituting the preliminary mapping (from step 2) into the corresponding
+   parameter types of the invocation target.  For each argument that is
+   recursively inferred, create additional constraints on type parameters using
+   the resulting static type.  Then proceed to step 5.
+
+4. Visit arguments (new): If experiment flag `inference-update-1` is enabled,
+   topologically sort the arguments into stages (see [topological
+   sort](#Topological-sort) below), and then for each stage, do the following:
+
+   * Recursively infer all arguments selected for this stage that are not
+     function literals (see [what constitutes a function
+     literal?](#what-constitutes-a-function-literal), in source order.  (Note
+     that an invariant of the topological sort is that non-function literals are
+     always placed in the first stage).  The logic for obtaining type contexts
+     and for creating additional constraints is the same as in step 3.
+
+   * Recursively infer all arguments selected for this stage that are function
+     literals, in source order.
```

Comment:
Heh, this is funny because an earlier (doc-based) version of the spec, I did not specify the order of processing function literal arguments, and Lasse suggested that we ought to specify it anyhow (even though it _shouldn't_ matter), because if there's any chance that we're wrong and it actually _does_ matter in some subtle way, we want to make sure the user sees consistent behavior.

I was swayed by that argument and decided to go ahead and specify the order.  I've added normative text explaining that we don't believe the order should matter, but we're specifying it anyway to reduce the risk of unpredictable behavior.

Reply:
Heh. ;-)